aesd-char-driver/main.c

```diff
@@ -163,11 +165,93 @@ ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
 	mutex_unlock(&device->char_device_lock);	
 	return retval;
 }
+
+loff_t aesd_llseek(struct file *filp, loff_t offset, int whence)
+{
+	loff_t ret_val;
+	struct aesd_dev *dev = filp->private_data;
+	loff_t size = dev->aesd_buffer.filled_buffer_size;
+	//include lock
+	if(mutex_lock_interruptible(&dev->char_device_lock))
+		return -ERESTARTSYS; 
+	ret_val = fixed_size_llseek(filp, offset, whence, size);
+	mutex_unlock(&dev->char_device_lock);
+	//remove lock
+	return ret_val;
+}
+
+long aesd_ioctl(struct file *filp, unsigned int cmd, unsigned long arg)
+{
+	long ret_val = 0;
+	struct aesd_seekto seekto;
+
+	switch(cmd)
+	{
+		case AESDCHAR_IOCSEEKTO:
+			{
+				if(copy_from_user(&seekto,(const void __user *)arg, sizeof(seekto))!=0)
+				{
+					ret_val = -EFAULT;
+				}
+				else
+				{
+					ret_val = aesd_adjust_file_offset(filp, seekto.write_cmd, seekto.write_cmd_offset);
+				}
+				break;
+			}
+		default:
+		{
+			ret_val = -EINVAL;
+		}
+	}
+	return ret_val;
+}
+
+static long aesd_adjust_file_offset(struct file *filp, unsigned int write_cmd, unsigned int write_cmd_offset)
+{
+	struct aesd_dev *dev = filp->private_data;
+	loff_t offset = 0;
+	int i = 0;
+	int ret_val =0;
+	if(!dev)
+	{
+		return -EINVAL;
+	}
+	if(mutex_lock_interruptible(&dev->char_device_lock))
+		return -ERESTARTSYS;
+	
+	if(write_cmd > (AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED-1))
+	{
+		printk(" Invalid write cmd");
+		return -1;
```

Comment:
@Ganeshkm10 what will happen to your lock in any of the error handling cases here?